### PR TITLE
Feature/weather app assessment

### PR DESCRIPTION
**Pull request by bpblattner**
Hello Contently team :)

I just finished the technical assessment in building a weather app that solves for:

As a user, I would like to search for the current weather in my area
As a user, I would like to see the 5-day forecast
As a user, I would like to share the URL of the weather in my area and see the results
As a user, I would like to toggle between celsius and fahrenheit

Notes:

I used tailwindcss framework to build out UI components
I created a container to call API and pull data, passing down to dumb UI components and rendered with data
I added clickable link allowing a user to toggle between C and F, with unique route for each
To solve for sharing the URL for an area with results I added unique routes by city and unit of measurement
The forecast API was tricky since it returns a much larger list than expected, with a new entry for every 3 hours on each day. The list returned the remaining hours of todays date as well, so I added logic to omit the entries from the forecast with todays date. To be honest, it would have been preferred to use their 16 day forecast api that doesnt give every 3 hours, but it was only available to paid users, so I took the first entry for each date not including todays.
I took the liberty of removing the rainbow colored background color from root in favor of a design that I felt fit better.
I enjoyed this challenge as there were some gotcha's with the weather API, and feedback is more than welcomed. I look forward to hearing back from you.

-Brian

### CODE REVIEW

**Comment by nethanelkohen** [comment]
Hey Brian, thanks for completing our take home assignment! I'm Nate from the engineering team at Contently. I'm going to drop a few comments in your PR. Looking forward to hearing back from you.

File: src/components/WeatherContainer.jsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  const toggleUnitType = () => {
+    let url;
+    if (props.match.params.unit === 'imperial') {
+      url = `/${city}/metric`
+    }
+    if (props.match.params.unit === 'metric') {
+      url = `/${city}/imperial`
+    }
+    history.push(url)
```
Comment (nethanelkohen) [inline src/components/WeatherContainer.jsx:41]: This is causing the DOM to re-render whenever someone changes the unit. Is there a reason why you chose this path? How would you refactor it so it doesn't cause a re-render?

File: src/components/WeatherContainer.jsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    window.location.reload(false);
```
Comment (nethanelkohen) [inline src/components/WeatherContainer.jsx:42]: This is causing the DOM to re-render whenever someone changes the unit. Is there a reason why you chose this path? How would you refactor it so it doesn't cause a re-render?

File: src/components/WeatherContainer.jsx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    fetchForecast(props.match.params.city, props.match.params.unit)
+    .then(results => 
+      results.list.forEach((i) => {
+        if (forecastDate !== i.dt_txt.split(" ")[0]) { //omit today, and since api returns every 3 hours for each day, take the first entry and skip the rest for that day
+          forecastDate = i.dt_txt.split(" ")[0]
+          setForecastData(oldArray => [...oldArray, {temp: i.main.temp.toFixed(), description: i.weather[0].description, image: setWeatherDataImage(i.weather[0].main), date: forecastDate}]);
+        }
+      })
+      )
+  }, [props.match.params.city])
```
Comment (nethanelkohen) [inline src/components/WeatherContainer.jsx:69]: Why did you use `useLayoutEffect` over `useEffect`?

File: src/components/WeatherContainer.jsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  useLayoutEffect(() => {
+    setCity(props.match.params.city);
+
+    if (props.match.params.unit === 'metric') {
+      setIsMetric(true)
+    } else {
+      setIsMetric(false)
+    }
+
+    fetchWeather(props.match.params.city, props.match.params.unit)
```
Comment (nethanelkohen) [inline src/components/WeatherContainer.jsx:54]: Do you foresee any pitfalls with using data from `props.match.params` instead of data stored in state? If so, let me know how you'd refactor your `fetchWeather` function with that in mind.

File: src/helpers/index.js [hunk 5]
```diff
@@ -0,0 +1 @@
+export * from './utils';
```
Comment (nethanelkohen) [inline src/helpers/index.js:1]: Could you explain the logic of including this file and exporting everything from `utils.js` as *?

File: src/helpers/utils.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    `http://api.openweathermap.org/data/2.5/forecast?q=${cityName}&units=${units}&appid=${API_KEY}`,
+  )
+    .then(response => response.json())
+    .then(data => {
+      return data;
+    })
+    .catch((error) => {
+      console.log(error)
+    })
+};
```
Comment (nethanelkohen) [inline src/helpers/utils.js:27]: How would you refactor this to handle errors on the UI? How would you send the error and what would you do with it?

**Comment by nethanelkohen** [comment]
Why did you decide against using state management? Even something simple like the Context API?

**Comment by nethanelkohen** [comment]
How would you improve the UX for users with 3G or slower connections?

**Comment by nethanelkohen** [comment]
When searching for West Park, NY, I am getting data for West Park, FL (coords returned are lat: 25.9845, lon: -80.1989). I know this is mainly an issue with the API, but how would you go about solving that issue?

**Comment by nethanelkohen** [comment]
Loving how clean and readable your code is! What is your favorite part of the code you wrote and what would you refactor if there weren't a time limit? Looking forward to hearing back! Thanks!

**Comment by bpblattner** [comment]
> Hey Brian, thanks for completing our take home assignment! I'm Nate from the engineering team at Contently. I'm going to drop a few comments in your PR. Looking forward to hearing back from you.

Hey @nethanelkohen, thank you so much for taking the time to review this thoroughly and quickly. I really appreciate the great questions and I'm glad to have the chance to explain each. I will start responding to each of these questions here, and send an email back to you when I finish the responses.

File: src/components/WeatherContainer.jsx (hunk 2 above)
Comment (bpblattner) [reply src/components/WeatherContainer.jsx:42]: This can be achieved by moving isMetric, and setIsMetric back one level to app.jsx and passing these down as props through the Route. I would make these key changes:

In app.jsx - 

```const App = () => {
  const [isMetric, setIsMetric] = useState(false);
  
  return (
    <BrowserRouter>
      <TopNav />
      <Switch>
        <Route exact path="/" component={Home} />
        <Route exact path="/:city/:unit" render={(props) => <WeatherContainer isMetric={isMetric} setIsMetric={setIsMetric} {...props} />} />
      </Switch>
    </BrowserRouter>
    );
}```

In WeatherContainer.jsx
  ```const { isMetric, setIsMetric } = props;

  const toggleUnitType = () => {
    let url;
    if (isMetric) {
      url = `/${city}/imperial`
      setIsMetric(false)
    } else {
      url = `/${city}/metric`
      setIsMetric(true)
    }

    history.push(url)
  }
```

File: src/components/WeatherContainer.jsx (hunk 4 above)
Comment (bpblattner) [reply src/components/WeatherContainer.jsx:54]: If the component being called from the route isn’t a direct child, then this value would come back undefined. This could be a pitfall in that if this was a real world client project, and new features were to be added and the project needed to scale, this scenario of an undefined value could play out if another engineer started building on this code base.

To solve for this, I could have added a new state value in App.jsx for this input (or even better Context as you mentioned), and capture the user input at the time its searched in TopNav.jsx, then passed down to the container 
```
render={(props) => <WeatherContainer SearchInput={SearchInput} {…props}/>}). 
```

In the FetchWeather function, I could add a conditional check to ensure that this value isn’t undefined before making the api call, and pull this this value from state, or even grab the value from the url params themselves using a library like query-string  — cityName ? cityName : fetchQueryStringValue(q). This would ensure this call is never made with an undefined value.

File: src/helpers/utils.js (hunk 6 above)
Comment (bpblattner) [reply src/helpers/utils.js:27]: I would return the error message from the util function and in WeatherContainer create a new state value for identifying there is an error, and passing the error message to the UI.

```
  if (response.ok) {
    return response;
  } else {
    throw new Error('Something went wrong. Please try your search again);
  }
})
    .then(response => response.json())
    .then(data => {
      return data;
    })
.catch((error) => {
      return error
    })
```

when rendering UI
```
{error && (
<AlertComponent alertMessage={error} />
```

File: src/helpers/index.js (hunk 5 above)
Comment (bpblattner) [reply src/helpers/index.js:1]: This wasn’t necessary at all and this file can be deleted as it serves no purpose. Since both functions in utils.js are exported there is no need for an index that exports *. I think at the time I was getting an error for the file not containing a default export, but I somehow didn’t notice that I had forgot to import each method within an object and my original import statement was incorrect. 

I think originally I had mistakenly imported this as 
Import fetchWeather from ‘../helpers/utils’;

And noticed it should be 

import { fetchWeather } from '../helpers/utils’

and then didn’t go back to delete the index.

File: src/components/WeatherContainer.jsx (hunk 3 above)
Comment (bpblattner) [reply src/components/WeatherContainer.jsx:69]: This one is a little embarrassing because I actually decided to go with useEffect as useLayoutEffect doesn’t apply here. It wasn’t until one of my final commits [(e60bb39faaf72a2461b4fdac6a8bb045e4ed0c65)](https://github.com/bpblattner/weather-app/pull/1/commits/e60bb39faaf72a2461b4fdac6a8bb045e4ed0c65) that I switched it from useEffect to useLayoutEffect because I was playing around with useLayoutEffect and seeing if it made a difference in how the page rendered. I actually didn’t intend to commit that change specifically and somehow missed it before sending over the PR >.<

 It’s funny you asked about steps I could take to improve performance on a 3G or slow connection because I’m currently on vacation in Mexico and was dealing with this exact problem and noticing the page was rendering slowly and was experimenting with ways to wait until the DOM was ready before the browser painted the UI. useLayoutEffect didn’t really apply though because I wasn’t changing a ref or anything related to the DOM and it wasn’t necessary to wait for that.

**Comment by bpblattner** [comment]
> Why did you decide against using state management? Even something simple like the Context API?

To be honest, when I first read the requirements for this challenge it seemed simple and small enough to not require any state management strategy outside of using a container with state values via hooks. In hindsight, you’re right in that Context API would’ve worked better as its fast and simple to setup and since I have the search input split into the TopNav component, on a separate level from the container and its children, Context would’ve made more sense. 

As I was reviewing your questions and reflecting on my approaches to this project and trying to finish it in time, I know now that I should've at LEAST moved some of these state values back from the container into app.jsx, OR created a container that was above TopNav.jsx and the UI components, OR went ahead and setup a context. I did run into issues with rendering between state changes and could've done better managing that. 

I was trying intentionally to keep it simple and clean but after I built this out it I recognize now that it would have been beneficial to create a context, and add the consumer at the App.jsx level for all of the children to share.

**Comment by bpblattner** [comment]
> How would you improve the UX for users with 3G or slower connections?

As mentioned in one of the above posts, I actually had this problem myself while working on this. I'm currently in Mexico and the internet is down at my airbnb, and I was hot spotting from my mobile device with a 3G connection while doing this assessment. 

Some things that come to mind would be:
1. Keep images local to project and compress them with something like squoosh.app
2. Bring in react-lazyload for images and components
3. If necessary introduce code-splitting using React features like dynamic imports with React.lazy

**Comment by bpblattner** [comment]
> When searching for West Park, NY, I am getting data for West Park, FL (coords returned are lat: 25.9845, lon: -80.1989). I know this is mainly an issue with the API, but how would you go about solving that issue?

This is a great question and actually something I was thinking about as I was reviewing the API docs. After looking a little deeper into this, it looks like OpenWeatherMaps allows you to search by city ID, which would be a lot more accurate. I also just found that they provide a nice minified json file that has a record for every city by its name, state, ID, and coords. 

If you were to search West Park, NY it looks like we could use this json to search against and match the city name, and state to the appropriate city ID and search that way instead of by name.

**Comment by bpblattner** [comment]
> Loving how clean and readable your code is! What is your favorite part of the code you wrote and what would you refactor if there weren't a time limit? Looking forward to hearing back! Thanks!

Thank you for the compliment, and thanks again for taking the time to review this. 

As you mentioned I'm pretty happy overall with how the code looks and its overall readability. I'm pretty happy as well with the overall design of the UI given the amount of time to create a design with no wireframes or reference point. I did try and keep it as simple and clean as possible without bringing in unnecessary 3rd party libraries or going over the top with design. 

If I had unlimited time I would have taken more time to ensure I was handling state in a better way, and handling some of the rendering issues the correct way. I would've also spent more time on handling the API responses with both error handling, and creating a user experience for when an error occurs, displaying that error, etc. I did notice when I ran this again last night that since the forecast API returns results on a rolling 3 hour basis, I was getting 6 days back when it excluded today, so I would've handled that. I would have also beat up the search a bit more and tested for things like two cities with the same name, and explored solutions around that. On that note, I would have written tests to catch edge cases like some of those mentioned.
